Upgrade to Gradle 5.0

alpar-t opened the pull request:
- Upgrade to Greadle 5.0 
- updates to comply with breaking changes in Groovy. 
It seems like the latest version brought along by Gradle changes how the `final` keyword is interpreted, 
unfortunately it can't see past branches so it doesn't work like ti would in java.

This is a work in progress to get the tests running. Will only merge when Gradle 5.0 is GA.

Release Notes: https://github.com/gradle/gradle/releases/tag/v5.0.0

alpar-t committed:
Upgrade to Gradle 5.0 Milestone 1

elasticmachine commented:
Pinging @elastic/es-core-infra

rjernst commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy:
```diff
@@ -563,7 +563,7 @@ class ClusterFormationTasks {
     }
 
     static Task configureInstallPluginTask(String name, Project project, Task setup, NodeInfo node, String pluginName, String prefix) {
-        final FileCollection pluginZip;
```
why is final removed?

atorok replied:
I will not merge the PR until there is a GA, just wanted to start running in CI. 

Gradle bumps the Groovy version, which now complains that the final variable is changed. 
It doesn't see past branches the same way java does. One more reason to switch to java.

rjernst reviewed (commented):
I have one question, but I also don't think we should upgrade to a milestone release. This is like upgrading to a beta.

alpar-t committed:
Merge branch 'master' into upgrade-gradle-5

alpar-t committed:
Add settings.gradle to unrelated project in the hierarchy

alpar-t committed:
Merge remote-tracking branch 'origin/master' into upgrade-gradle-5

alpar-t committed:
Merge branch 'master' into upgrade-gradle-5

alpar-t committed:
Merge branch 'master' into upgrade-gradle-5

alpar-t committed:
Merge remote-tracking branch 'origin/master' into upgrade-gradle-5

alpar-t committed:
Upgrade to Grale 5.0 RC1

alpar-t committed:
Merge remote-tracking branch 'origin/master' into upgrade-gradle-5

alpar-t committed:
remove more finals

alpar-t committed:
fix annotation processor

alpar-t committed:
Fix ordering

Looks like Gradle changed the order in which things happen here.
There were no guarantees on ordering before, the order just happened to
be  right.

alpar-t committed:
Merge remote-tracking branch 'origin/master' into upgrade-gradle-5

alpar-t committed:
Merge remote-tracking branch 'origin/master' into upgrade-gradle-5

alpar-t committed:
Merge remote-tracking branch 'origin/master' into upgrade-gradle-5

alpar-t committed:
bump scan plugin version

alpar-t committed:
Upgrade to 5.0 RC3

alpar-t committed:
remove feature preview, the feature is here

alpar-t committed:
Merge remote-tracking branch 'origin/master' into upgrade-gradle-5

atorok commented:
@elasticmachine test this please

alpar-t committed:
Upgrade to RC 4

alpar-t committed:
Merge remote-tracking branch 'origin/master' into upgrade-gradle-5

alpar-t committed:
Merge remote-tracking branch 'origin/master' into upgrade-gradle-5

alpar-t committed:
Merge remote-tracking branch 'origin/master' into upgrade-gradle-5

alpar-t committed:
fix test integ task

alpar-t committed:
Merge remote-tracking branch 'origin/master' into upgrade-gradle-5

alpar-t committed:
Merge remote-tracking branch 'origin/master' into upgrade-gradle-5

alpar-t committed:
Upgrade to 5.0 GA

alpar-t committed:
remove redundant copy

alpar-t committed:
Merge remote-tracking branch 'origin/master' into upgrade-gradle-5

alpar-t committed:
Fix for copying YML tests

alpar-t committed:
Deal with deprecations

alpar-t committed:
Resolve the configuration before copying

atorok commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/test/RestIntegTestTask.groovy:
```diff
@@ -215,7 +215,7 @@ public class RestIntegTestTask extends DefaultTask {
      * @param includePackagedTests true if the packaged tests should be copied, false otherwise
      */
-    static Task createCopyRestSpecTask(Project project, Provider<Boolean> includePackagedTests) {
+    Task createCopyRestSpecTask() {
```
The static method was called with the value of `includePackagedTests` in what happened to be the right order so that changes to the property were picked up in time and the right value was passed into the `project.afterEvaluate` ( which had no use, since the value was already picked up anyhow ). 
This can't be static to make changes to the property reflected as they should. Gradle 5 seems to have changed the order in which some operations are executed.  I wish they would have an option to randomize ordering of otherwise undefined sequences.

alpar-t committed:
Add dependency to resolve configurations

atorok commented:
@rjernst Gradle 5.0 is GA and this is ready for review and merge now.

droberts195 commented:
One thing to note is that in release manager Gradle will run with ml-cpp and found-elasticsearch-plugin in the elasticsearch-extra directory.  Were there any particular things that needed changing as part of this PR that are also likely to need changing in the `build.gradle` files in those repos before Gradle 5 will work with them?  I'm happy to make the ml-cpp changes if any are required, but it would be good to consider this before merging this PR because otherwise the unified snapshot will be broken in the meantime.

atorok commented:
@droberts195 The best thing is to run with a build scan and look at deprecation messages. 
The `finally` keyword changed how it works and I had to remove it from a bunch of places. 
Other than that this was not special  from minor version updates where things are ran in different order and this uncovers some explicit ordering that is missing.  I'm also glad to help making those changes. I think the hadoop plugin is also affected ( /cc @jbaiera )

droberts195 commented:
I can confirm that ml-cpp now works with Gradle 5

rjernst reviewed (approved):
LGTM

atorok commented:
`gradlew` in the found plugin actually uses the one from the ES repo. I checked to make sure Gradle 5 doesn't cause any issues there.
